File: frontend/src/api/notifications.ts
/**
 * Notifications API client
 */
import api from './client';

export interface Notification {
    id: number;
    user_id: number;
    type: 'deadline_reminder' | 'status_change' | 'visa_reminder' | 'weekly_digest' | 'community_reply' | 'community_like';
    priority: 'critical' | 'high' | 'medium' | 'low';
    title: string;
    message: string;
    link?: string;
    is_read: boolean;
    created_at: string;
    read_at?: string;
}

export interface NotificationListResponse {
    notifications: Notification[];
    total: number;
    unread_count: number;
}

export interface NotificationPreferences {
    email_deadline_reminders: boolean;
    email_status_changes: boolean;
    email_visa_reminders: boolean;
    email_weekly_digest: boolean;
    inapp_deadline_reminders: boolean;
    inapp_status_changes: boolean;
    inapp_visa_reminders: boolean;
    inapp_weekly_digest: boolean;
}

export const notificationsApi = {
    /**
     * Get paginated list of notifications
     */
    getNotifications: async (
        limit = 20,
        offset = 0,
        unreadOnly = false
    ): Promise<NotificationListResponse> => {
        const response = await api.get('/notifications', {
            params: { limit, offset, unread_only: unreadOnly },
        });
        return response.data;
    },

    /**
     * Get unread notification count
     */
    getUnreadCount: async (): Promise<{ unread_count: number }> => {
        const response = await api.get('/notifications/unread-count');
        return response.data;
    },

    /**
     * Mark a single notification as read
     */
    markAsRead: async (notificationId: number): Promise<void> => {
        await api.put(`/notifications/${notificationId}/read`);
    },

    /**
     * Mark all notifications as read
     */
    markAllAsRead: async (): Promise<{ message: string }> => {
        const response = await api.put('/notifications/mark-all-read');
        return response.data;
    },

    /**
     * Get notification preferences
     */
    getPreferences: async (): Promise<NotificationPreferences> => {
        const response = await api.get('/notifications/preferences');
        return response.data;
    },

    /**
     * Update notification preferences
     */
    updatePreferences: async (
        preferences: Partial<NotificationPreferences>
    ): Promise<NotificationPreferences> => {
        const response = await api.put('/notifications/preferences', preferences);
        return response.data;
    },
};

File: backend/app/routers/notifications.py
"""
Notifications API routes
"""
from typing import Optional, List
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session

from app.database import get_db
from app.auth import get_current_user
from app.models.user import User
from app.models.notification import Notification, UserNotificationPreference
from app.schemas.notification import (
    NotificationResponse, NotificationListResponse, UnreadCountResponse,
    MarkReadRequest, NotificationPreferenceResponse, NotificationPreferenceUpdate
)
from app.services.notification_service import get_notification_service

router = APIRouter(prefix="/notifications", tags=["Notifications"])


# ============== Notification Endpoints ==============

@router.get("", response_model=NotificationListResponse)
async def get_notifications(
    limit: int = Query(20, ge=1, le=100),
    offset: int = Query(0, ge=0),
    unread_only: bool = Query(False),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get user notifications with pagination"""
    service = get_notification_service()
    notifications, total, unread = service.get_notifications(
        db=db,
        user_id=current_user.id,
        limit=limit,
        offset=offset,
        unread_only=unread_only
    )
    
    return NotificationListResponse(
        notifications=[NotificationResponse.model_validate(n) for n in notifications],
        total=total,
        unread_count=unread
    )


@router.get("/unread-count", response_model=UnreadCountResponse)
async def get_unread_count(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get unread notification count for badge"""
    service = get_notification_service()
    count = service.get_unread_count(db, current_user.id)
    return UnreadCountResponse(unread_count=count)


@router.put("/{notification_id}/read")
async def mark_notification_read(
    notification_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Mark a single notification as read"""
    notification = db.query(Notification).filter(
        Notification.id == notification_id,
        Notification.user_id == current_user.id
    ).first()
    
    if not notification:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Notification not found"
        )
    
    service = get_notification_service()
    service.mark_as_read(db, current_user.id, [notification_id])
    
    return {"message": "Notification marked as read"}


@router.put("/mark-all-read")
async def mark_all_read(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Mark all notifications as read"""
    service = get_notification_service()
    updated = service.mark_as_read(db, current_user.id)
    return {"message": f"{updated} notifications marked as read"}


# ============== Preference Endpoints ==============

@router.get("/preferences", response_model=NotificationPreferenceResponse)
async def get_preferences(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get user notification preferences"""
    service = get_notification_service()
    preferences = service.get_or_create_preferences(db, current_user.id)
    return NotificationPreferenceResponse.model_validate(preferences)


@router.put("/preferences", response_model=NotificationPreferenceResponse)
async def update_preferences(
    updates: NotificationPreferenceUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Update user notification preferences"""
    service = get_notification_service()
    preferences = service.get_or_create_preferences(db, current_user.id)
    
    # Update only provided fields
    update_data = updates.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        if value is not None:
            setattr(preferences, field, value)
    
    db.commit()
    db.refresh(preferences)
    
    return NotificationPreferenceResponse.model_validate(preferences)


# ============== Admin/Debug Endpoints ==============

@router.post("/test/deadline-reminders", include_in_schema=False)
async def test_deadline_reminders(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """[Dev Only] Manually trigger deadline reminder check"""
    if current_user.role.value != "admin":
        raise HTTPException(status_code=403, detail="Admin only")
    
    service = get_notification_service()
    count = service.send_deadline_reminders(db)
    return {"message": f"Sent {count} deadline reminders"}


@router.post("/test/visa-reminders", include_in_schema=False)
async def test_visa_reminders(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """[Dev Only] Manually trigger visa reminder check"""
    if current_user.role.value != "admin":
        raise HTTPException(status_code=403, detail="Admin only")
    
    service = get_notification_service()
    count = service.send_visa_reminders(db)
    return {"message": f"Sent {count} visa reminders"}


@router.post("/test/weekly-digest", include_in_schema=False)
async def test_weekly_digest(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """[Dev Only] Manually trigger weekly digest"""
    if current_user.role.value != "admin":
        raise HTTPException(status_code=403, detail="Admin only")
    
    service = get_notification_service()
    count = service.send_weekly_digests(db)
    return {"message": f"Sent {count} weekly digests"}
